### PR TITLE
8822 incomplete datasets via api

### LINKED ISSUE #8822: Feature Request/Idea: Possibility to create & save dataset with incomplete mandatory metadata via API only
**Overview of the Feature Request**
It would be possible to create and save a dataset via API without all the mandatory metadata being present. This would only be possible via API to ensure that the user interface remains the same and doesn't get more complicated. It would be a way to make it easier to connect/push in data from other active data platforms.
Because it's only via API, the user interface in Dataverse can remain the same and would be perfect already for a user to go through to  fill in their missing metadata before publication. The only changes would maybe be some more notifications or tags where necessary to point out that the metadata needs to be completed before publication is possible.

**What kind of user is the feature intended for?**
API User

**What inspired the request?**
We want to make it easier for users to create a dataset using data that is present in another application that doesn't necessarily contain all the mandatory metadata options or flexibility needed for a Dataverse dataset.

**What existing behavior do you want changed?**
The create & save a dataset action via API.

**Any brand new behavior do you want to add to Dataverse?**
The possibility to save a dataset's metadata without all the mandatory metadata filled in via API only.

**Any related open or closed issues to this feature request?**
No, but we (KU Leuven LIBIS) would definitely be willing to work on this and develop it. We just first want to gauge the interest/willingness and get input from the community on this feature.

### PR DESCRIPTION
**Pull request by ErykKul**
**What this PR does / why we need it**:
When integrating GitLab or iRods with Dataverse, we want researchers to be able to send files and just enough information to get the dataset created via API. Publishing is prevented until required fields are entered (using the GUI, most often).

**Which issue(s) this PR closes**:

Closes #8822

**Suggestions on how to test this**:
First, you will need to update the solar configuration. For this feature to work, a new solr field is required: "datasetValid". The needed configuration is added in "schema.xml". Notice that when you do not use this feature, updating solr configuration and/or reindexing is not required. Nevertheless, it is reasonable to keep the solr configuration up-to-date and do the configuration update and reindex.

Next, you need to enable these settings:
- :AllowInvalidMetadataThroughAPI: set this to "true". If you do that without updating the solr configuration first, you will not be able to ingest dataset anymore due to the missing field errors.
- :CanReviewInvalid: can be set optionally to "true", you can then submit datasets with invalid metadata for review. Otherwise, you will need to correct the metadata before sending to review. Datasets with invalid metadata can never be published, regardless of any setting.
- :ShowValidityFilter: when set to "true" it enables a filter in "My Data" page (important: this filter will only work correctly when all metadata is reindex on the solr server, otherwise valid datasets will not be shown):

image

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
When the new feature is not used, there are no changes to the user interface. When you use the new feature, you will have the possibility to enable filtering of valid and/or invalid metadata on the "My Data" page, as discussed earlier.

Invalid metadata is marked with a new tag "Invalid metadata":

image

This tag is visible in the "My Data" page and the main page, when logged in. When you do not log in, you can only see published metadata, and since invalid metadata cannot be published, you will not see that tag. The tag, together with a warning, is also shown when viewing a dataset with invalid metadata:

image

A warning is also shown when submitting for review and trying to publish a dataset with invalid metadata:

image

The exact content of the tag and the text of the warning can be modified with the "notValid" and "dataset.message.invalid.warning" properties, respectively.

**Is there a release notes update needed for this change?**:
Yes.

**Additional documentation**:
You can also watch the demo as given during the community meeting. You can find the video at the top of DataverseTV: https://dataverse.org/dataversetv

Here's a direct link: https://harvard.zoom.us/rec/share/EZGLPBLPv3o74H_WD-ejp34c3Grro1Jk8mS5L8um9PyFbKzVr2Ro_62gqIkckKf5.1Fh8UfNOIkc8sqpc

### CODE REVIEW

**Comment by pdurbin** [comment]
Related:

- #7963

**Comment by DieuwertjeBloemen** [comment]
For a bit of context: at KU Leuven RDR, we’re working on an integrations dashboard to easily pull data from exterior data management sources such as our university’s iRods installation, Github, Gitlab and hopefully RedCap and OSF in the near future to name a few.
What makes it different from other Dataverse integrations is that we're not trying to push data from a system to Dataverse, but to pull from a system into Dataverse. This way we have more control over the integration and it allows for more integrations with other data systems as we don't need the system's coöperation, only an API interface that allows for pulling data etc.

So, we're working on an integrations dashboard that alllows our users to pull data from external data sources such as Github, Gitlab, iRods, RedCap, OSF, etc.
Our researchers would currently be able to add data to existing draft datasets from those sources without having to first download the files on their computer, but we would like to add the option to pull files and immediately create a new draft dataset as well. This way they don't have to keep going back and forth. They can immediatly go to the integrations dashboard to select and pull the data and are afterwards sent to Dataverse to supply the metadata for the new dataset. It made more sense not to have them fill in the metadata in the integrations dashboard, as we would have to create a UI for that, and also keep it in line with our Dataverse fields and guidance. So, it made more sense to let them create a dataset without all the metadata.

I've added my attempt at putting our 'integrations dashboard' into a rudimentary flowchart to give it a bit more context. (I might correct it if my colleagues point out any errors). The part where this PR is necessary for is highlighted in yellow.

Draft integrations dashboard KU Leuven RDR 20230105

**Comment by mreekie** [comment]
Prio meeting
- Notes discussed at tech hours.
- There is some existing work added in DV4 having to do with migrating data
- This functionality involves API and UI changes.  The team would typically get input from the product owner and the UI person.
- There design decisions to make, that go beyond creating a new function.
- We can possibly look to Julian here.

Next steps:
- For now we are creating a new lable: UIUX: Design that indicates that input is required from the product owner and on the design of the user experience.

**Comment by DieuwertjeBloemen** [comment]
@pdurbin Thanks for the update. If there's anything we can do to help, be sure to let us know.

BTW, we were thinking of possibly doing a demo of the UI etc. of our 'integrations dashboard' in the community call on February 7th. We're currently working on the documentation and on making it as flexible as possible so others at some point might be able to use it for their own set-up as well. I'll send an email once we're sure that we're ready to show it by then to ask if there's time for it.

**Comment by DieuwertjeBloemen** [comment]
@pdurbin Hi,
I had another look and think about the points you noted as needing consideration and here is some additional info on them:

1. _alternative phrasing for "invalid metadata" (maybe "missing")_
o	Some ideas: "incomplete metadata" , "missing metadata", "unfinished metadata", "lacking metadata", "failed metadata validation", "unvalidated metadata", 
o	We ended up with ‘invalid’ because that is what it technically is. There could possibly be metadata, but is just hasn’t been validated. But we’re definitely open to other options as well. It’s something we’ve gone back and forth on ourselves as well, because ‘invalid’ does sound a bit intimidating and maybe doesn't make it super clear what the user is supposed to do with this information. Maybe the "lacking metadata" is the best one semantically, as it can mean both missing and invalid. But I personally don't like the sound of it ;)

2. _explore relationship with this issue (for create or publish): #Query Dataverse for mandatory metadata fields via API Query Dataverse for mandatory metadata fields via API #6978_ 
o	It might be the case that this API call already solves that issue or at least the underlying question. This is because this API call doesn’t have the condition that there is no metadata, it just has the condition that the metadata isn’t validated to also allow datasets without any metadata. (@ErykKul : did I get that correctly?) So, you can create a dataset without all the mandatory metadata present using this call, and then complete and possibly correct them via the DV UI. But maybe that is not covering the entire need of the use case.

Let us know if there is anything we need to take a look at or that we can further help out with.

File: doc/sphinx-guides/source/api/native-api.rst [hunk 1]
```diff
@@ -526,6 +526,39 @@ To create a dataset, you must supply a JSON file that contains at least the foll
 - Description Text
 - Subject
 
+Alternatively, you can turn off the validation of the dataset by using the optional "doNotValidate" parameter. This parameter works only when the ":AllowInvalidMetadataThroughAPI" setting is set to "ture". Your solr configuration must be up-to-date with the "schema.xml" file for this feature to work.
```
Comment (sekmiller) [inline doc/sphinx-guides/source/api/native-api.rst:529]: typo here "ture"

Comment (ErykKul) [reply doc/sphinx-guides/source/api/native-api.rst]: Thanks! Typo is fixed. I have also merged the current develop branch into this pull request branch.

**Comment by sekmiller** [comment]
Suggestions from Tech discussion:

Change wording of invalid data to "incomplete" in bundle.
Gray out Publish Button when dataset version is in incomplete state.
recommend use of MPConfig.
Reconsider whether we want to make this part of core code since it's not likely to be implemented widely - (check with Sonia, et al to verify this)

**Comment by sekmiller** [comment]
Reviewed the request with Sonia. She thinks this is OK to implement since there will still be validations run prior to publication and noted that there are some current users who would just post "dummy data" to pass validation and maybe not revise before publication.

**Comment by mreekie** [comment]
sprint kickoff:
- moving this down to a 3.
- waiting on eryk's changes and then it willbe reviewed.

**Comment by ErykKul** [comment]
@sekmiller 
I have changed the wording. I have also disabled the publish button by adding the "is valid" requirement to the showPublishLink logic:
```
<ui:param name="showPublishLink" value="#{DatasetPage.workingVersion == DatasetPage.dataset.latestVersion 
                                                      and DatasetPage.dataset.latestVersion.versionState=='DRAFT'
                                                      and permissionsWrapper.canIssuePublishDatasetCommand(DatasetPage.dataset)
                                                      and DatasetPage.valid}"/>
```

I am not sure about the microprofile config. I thought that sooner or later all settings would be configurable in mp config? Since we check the settings at one place, i.e., the `SettingsServiceBean.java`, it would be ease to implement that all settings could be overwritten in the mp config, for example, by setting "mp.config.settings.:Key" value in mp config. Is it something that we would do in the future? Or we will have to "migrate" the configuration?

**Comment by pdurbin** [comment]
> I am not sure about the microprofile config. I thought that sooner or later all settings would be configurable in mp config? Since we check the settings at one place, i.e., the `SettingsServiceBean.java`, it would be ease to implement that all settings could be overwritten in the mp config, for example, by setting "mp.config.settings.:Key" value in mp config. Is it something that we would do in the future? Or we will have to "migrate" the configuration?

@poikilotherm what do you think? Do you want @ErykKul to switch the three database settings in this PR to MPCONFIG?

**Comment by pdurbin** [comment]
> **Guides Build Status / docs (pull_request) ** Failing after 24s

@ErykKul `make html` is failing for the guides. Can you please fix this? If you'd like to try our our fancy new Docker-based way to build the guides: https://preview.guides.gdcc.io/en/develop/developers/documentation.html#building-the-guides-with-a-sphinx-docker-container

**Comment by poikilotherm** [comment]
Regarding MPCONFIG.

*TL;DR: Static nature? Doesn't need flip in API test? Go MPC.*

First: I do see the necessity of DB settings for stuff that is not of a static nature. Things that change regularly. Things that otherwise would be hard to do.

If I get it right, the DB settings have been introduced in the first place to allow changing settings during testing via API and for UI related things that folks wanted to change without restarting the server.
These are historically grown choices, that are long time outdated. Nowadays you even can change system properties via asadmin at runtime. So more or less we just got maybe a bit stuck.

There is a vision in my head to refurbish the whole DB settings things because when this was created, a whole bunch of other technical debt was introduced. This refactoring would also allow to make the settings service ask the JvmSettings wrapper for values if nothing is within the database.

However, this is not a reality as of yet and requires another larger undertaking of code work. By the way, when you access settings from the UI code, you shall use the SettingsWrapper to avoid unnecessary DB queries on lots of page loads. (I did not look if you did already. And yet again, we must use this spring of tech debt that is there for reasons that were perfectly right at the time they were made.)

So, to return to the matter at hand and whether or not these three settings should "to DB or not to DB", I'd ask for now: do you expect this setting to remain (rather) static in a production use case once a choice has been made? If so, I'd vote for using a JvmSetting (which is, btw, making it much easier to allow only certain values and have verification during startup at a later point.)

But: right now, I did not yet refactoring our API tests to be JUnit 5 and have a nice helper to switch JvmSettings in a remote fashion. So if that is necessary right now, maybe one needs a DB setting ATM. (Watch the tears in my eyes water 'em. So much to do, so little time.)

Hope this helps.

**Comment by pdurbin** [comment]
@poikilotherm yes, this helps a lot! 😅 

@ErykKul to highlight Oliver's question: `do you expect this setting to remain (rather) static in a production use case once a choice has been made?`

**Comment by poikilotherm** [comment]
Hey I added a TL;DR! 😛

File: src/main/webapp/dataset.xhtml [hunk 2]
```diff
@@ -23,7 +23,8 @@
             <ui:param name="showMessagePanel" value="#{true}"/>
             <ui:param name="showPublishLink" value="#{DatasetPage.workingVersion == DatasetPage.dataset.latestVersion 
                                                       and DatasetPage.dataset.latestVersion.versionState=='DRAFT'
-                                                      and permissionsWrapper.canIssuePublishDatasetCommand(DatasetPage.dataset)}"/>
+                                                      and permissionsWrapper.canIssuePublishDatasetCommand(DatasetPage.dataset)
+                                                      and DatasetPage.valid}"/>
```
Comment (sekmiller) [inline src/main/webapp/dataset.xhtml:27]: @ErykKul  Sorry about the confusion, but when I requested "graying out" the publish button, I was really asking for it to be disabled. We want to show it but the logic for disabling it is around 319 and a bit beyond.

**Comment by sekmiller** [comment]
I fixed the Guides issue. It was just complaining that the underline for some of the settings were the wrong length

**Comment by pdurbin** [comment]
@poikilotherm offered to help move the database setting to MPCONFIG (regular, not feature flags).

@ErykKul meanwhile, can you please make the "grey out the Publish button" @sekmiller mentioned above?

**Comment by ErykKul** [comment]
Thanks for the help. It has been a very busy time lately for us, we just updated the production with a patched version to allow for the datasets with thousands of files. We needed to increase the heap space, other than that, everything looks good now.
I have just changed the logic to disabling i.s.o. hiding the publish button when metadata is not valid.
I will go through the documentation and some code to see how to change the settings to mp config.

**Comment by ErykKul** [comment]
I have tried implementing the validation settings via microprofile. I think it is OK now.

**Review by sekmiller — APPROVED** [review]
Thanks for this. sorry about all of the delays

**Comment by poikilotherm** [comment]
Sorry, I dragged this out of review to do some minor cleanup. Thx for bearing with us and keeping up working on this @ErykKul !

**Review by poikilotherm — APPROVED** [review]
I restructured the JvmSettings a bit, extended the docs and also added an Info API endpoint, so a client may detect if sending incomplete data will be acceptable before trying it.

Looks good to me, off to QA :-) 

(One thing though: I'm missing tests for this... There has been no additional API test added with deactivated validation)

@kcondon I see sometimes the CI pipeline for Maven Unit Tests fail - the Test Coverage submission seems to be flaky, tests run OK.

**Comment by ErykKul** [comment]
@poikilotherm 
Great changes! Thanks!

**Comment by kcondon** [comment]
Minor points:
1. dataverse.api.allow-incomplete-metadata section links to 
https://guides.dataverse.org/en/8822_incomplete_datasets_via_api/api/native-api.html#create-dataverse-api 
but should probably link to 
https://guides.dataverse.org/en/8822_incomplete_datasets_via_api/api/native-api.html#create-a-dataset-in-a-dataverse-collection

2. dataverse.ui.show-validity-filter section points to MyData page in guides, 
When enabled, the filter for validity of metadata is shown in:
https://guides.dataverse.org/en/8822_incomplete_datasets_via_api/user/account.html#my-data
but it doesn't show the validity filter enabled on the MyData page.

**Comment by ErykKul** [comment]
@kcondon 
I have addressed the doc changes you requested. Can you review the latest commits?

I have added a small paragraph in the release notes about the properties that can be set. Would that be sufficient for the admins to configure the feature?

For the edge case(s), I did the following: when a dataset has no title, its global ID would be used as the display name. This would improve linking and displaying information in all places for datasets without a title (also fixing the notifications links).

**Comment by kcondon** [comment]
@ErykKul Thanks for addressing all of the issues!
